File: Design and Programming Environments/Iteration 3/TicketManager/src/main/java/Service/UserService.java
package Service;
import Domain.User;
import Repository.UserRepository;
import Utils.DataChanged;
import Utils.Observable;
import Utils.Observer;

import java.util.ArrayList;
import java.util.List;

public class UserService implements Observable<DataChanged> {

    private List<Observer<DataChanged>> observers=new ArrayList<>();
    private UserRepository userRepository;


    public UserService(UserRepository userRepository){
        this.userRepository=userRepository;
    }

    public void save(String username,String password){
        User user=new User(username,password);
        userRepository.save(user);
    }

    public void delete(Integer integer){
        userRepository.delete(integer);
    }

    public void update(Integer integer,String username,String password){
        User user=new User(username,password);
        userRepository.update(integer,user);
    }

    public User findOne(String username){
        return userRepository.findOne(username);
    }

    public Iterable<User> findAll(){
        return userRepository.findAll();
    }
    @Override
    public void addObserver(Observer<DataChanged> e) {
        observers.add(e);
    }

    @Override
    public void removeObserver(Observer<DataChanged> e) {
        observers.remove(e);
    }

    @Override
    public void notifyObservers(DataChanged t) {
        observers.forEach(o->o.update(t));
    }

    public boolean checkPassword(String username, String password) {
        User user = userRepository.findOne(username);

        if (user!=null) {
            String userPassword=user.getPassword();
            return userPassword.equals(password);
        }
        return false;
    }
}

File: Second year/Software systems engineering/LibraryApplication/Code/LibraryApplication/LibraryApplication/Service/UserService.cs
using LibraryApplication.Domain;
using LibraryApplication.Repository;
using LibraryApplication.Utils;
using LibraryApplication.Validator;
using System.Collections.Generic;

namespace LibraryApplication.Service
{
    public class UserService : IObservable<IEvent>
    {
        public readonly UserRepository userRepository;
        public readonly BookRepository bookRepository;
        public readonly RentalRepository rentalRepository;
        public readonly CopyRepository copyRepository;
        public List<IObserver<IEvent>> Observers;

        public UserService(UserRepository userRepository, BookRepository bookRepository, CopyRepository copyRepository, RentalRepository rentalRepository)
        {
            this.userRepository = userRepository;
            this.bookRepository = bookRepository;
            this.copyRepository = copyRepository;
            this.rentalRepository = rentalRepository;
            Observers = new List<IObserver<IEvent>>();
        }

        public User LogIn(string username, string password)
        {
            User user = null;
            foreach (User u in userRepository.GetAll())
            {
                if (u.Id == username && u.Password == password)
                {
                    user = u;
                }
            }
            if (user == null)
                throw new ValidationException("Username or password incorrect! \n");
            return user;
        }

        public void AddObserver(IObserver<IEvent> e)
        {
            Observers.Add(e);
        }

        public void NotifyObservers(IEvent t)
        {
            Observers.ForEach(x => x.Update(t));
        }

        public void RemoveObserver(IObserver<IEvent> e)
        {
            Observers.Remove(e);
        }
    }
}